### PR TITLE
test: remaining branches coverage in per-diem page

### PR DESCRIPTION
**Pull request by suyashpatil78**
### <samp>🤖 Generated by Copilot at 7b60665</samp>

This pull request improves the code quality and the test coverage of the `AddEditPerDiem` module, which allows users to create or edit per diem expenses in the Fyle mobile app. It adds or updates unit tests for the components that handle the different steps of the per diem workflow, and it removes or simplifies some redundant or unnecessary code in the main component.

### <samp>🤖 Generated by Copilot at 7b60665</samp>

> _`PerDiem` pages_
> _Testing and refactoring_
> _Autumn leaves fall clean_

### <samp>🤖 Generated by Copilot at 7b60665</samp>

*  Remove unused and redundant code from `AddEditPerDiemPage` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dL296-L299), [link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dL2331-L2343))
*  Remove unnecessary and inconsistent condition from `setDefaultValues` function in `AddEditPerDiemPage` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dL553-L560))
*  Remove optional chaining operators from expressions that would always have the properties in `AddEditPerDiemPage` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dL607-R595), [link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dL813-R801), [link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-b6e1f6e315ef2edc53799d40c1304deff11e97673e1003986d199937196dda0dL1260-R1248))
*  Import `Subject`, `Subscription` and `finalize` modules from `rxjs` to `AddEditPerDiem5Page` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-cdcacda6d8616f7b0cbf830f18175b69dccaf827ea24d56f98bebe7fd374d4f6L34-R34))
*  Import mock data for testing `getPolicyDetails` function in `AddEditPerDiem5Page` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-cdcacda6d8616f7b0cbf830f18175b69dccaf827ea24d56f98bebe7fd374d4f6R43))
*  Add test cases for `getPolicyDetails` and `ionViewWillLeave` functions in `AddEditPerDiem5Page` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-cdcacda6d8616f7b0cbf830f18175b69dccaf827ea24d56f98bebe7fd374d4f6R615-R661))
*  Add test case for `isPaymentModeValid` function in `AddEditPerDiem3Page` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-1bbe20cca39fb0a9144847548b98cfc6e19b5ec063ff56c3d6d2ea7310a31ed3R194-R205))
*  Add test cases for `individualProjectIds$` and `isBalanceAvailableInAnyAdvanceAccount$` observables in `AddEditPerDiem2Page` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-e07d5851dc6ce6b9d4929a5a7cd5cce88c19fb320d9efbd769f8f1911bbd440dR595-R607), [link](https://github.com/fylein/fyle-mobile-app/pull/2336/files?diff=unified&w=0#diff-e07d5851dc6ce6b9d4929a5a7cd5cce88c19fb320d9efbd769f8f1911bbd440dR764-R789))

## Clickup
https://app.clickup.com/t/85ztk93f4

## Code Coverage
Please add code coverage here

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts [hunk 1]
```diff
@@ -293,10 +293,6 @@ export class AddEditPerDiemPage implements OnInit {
     );
   }
 
-  get showSaveAndNext(): boolean {
```
Comment (suyashpatil78) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts:296]: this method is not used anymore

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts [hunk 2]
```diff
@@ -550,14 +546,6 @@ export class AddEditPerDiemPage implements OnInit {
           const control = keyToControlMap[defaultValueColumn];
           if (!control.value && defaultValueColumn !== 'billable') {
             control.patchValue(defaultValues[defaultValueColumn]);
-          } else if (
```
Comment (suyashpatil78) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts:553]: this is unreachable part of the code as we are already checking conditions above, removing this

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts [hunk 3]
```diff
@@ -810,7 +798,7 @@ export class AddEditPerDiemPage implements OnInit {
         if (formValue.paymentMode?.acc?.type === AccountType.ADVANCE) {
           if (
             etxn?.tx.id &&
-            formValue.paymentMode?.acc?.id === etxn.tx.source_account_id &&
```
Comment (suyashpatil78) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts:813]: we already checked this conditions in line 798, therefore removing this

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts [hunk 4]
```diff
@@ -1257,8 +1245,7 @@ export class AddEditPerDiemPage implements OnInit {
               map(
                 (accounts) =>
                   accounts.filter(
-                    (account) =>
-                      account?.acc?.type === AccountType.ADVANCE && account?.acc?.tentative_balance_amount > 0,
+                    (account) => account?.acc?.type === AccountType.ADVANCE && account.acc.tentative_balance_amount > 0,
```
Comment (suyashpatil78) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts:1248]: already checked if acc and account are defined or not, removing chaining from statement after `&&`

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts [hunk 5]
```diff
@@ -2328,19 +2315,6 @@ export class AddEditPerDiemPage implements OnInit {
     }
   }
 
-  scrollCommentsIntoView(): void {
```
Comment (suyashpatil78) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem.page.ts:2331]: not used anymore

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts [hunk 6]
```diff
@@ -748,6 +761,32 @@ export function TestCases2(getTestBed) {
         });
       });
 
+      it('should set isBalanceAvailableInAnyAdvanceAccount$ correctly if paymentMode changed and paymentMode.acc.type equals PERSONAL account and account.acc is undefined', (done) => {
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts:764]: What is `correctly`? Should it be set to true, or false?

File: src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts [hunk 7]
```diff
@@ -748,6 +761,32 @@ export function TestCases2(getTestBed) {
...
+        accountsService.getEMyAccounts.and.returnValue(of(mockAccountsData));
+        component.ionViewWillEnter();
+        component.isBalanceAvailableInAnyAdvanceAccount$.subscribe((res) => {
+          expect(accountsService.getEMyAccounts).toHaveBeenCalledTimes(1);
+          expect(res).toBeTrue();
+          done();
+        });
+      });
+
+      it('should set isBalanceAvailableInAnyAdvanceAccount$ correctly if paymentMode changed and paymentMode.acc.type equals PERSONAL account and account is undefined', (done) => {
```
Comment (Chethan-Fyle) [inline src/app/fyle/add-edit-per-diem/add-edit-per-diem-2.page.spec.ts:777]: Same as above.

**Review by Chethan-Fyle — APPROVED** [review]

**Review by Dimple16 — APPROVED** [review]

**Pull request merged**